pyro/contrib/easyguide/easyguide.py

```diff
@@ -0,0 +1,267 @@
+            self.plates[name] = pyro.plate(name, size, subsample_size, subsample, *args, **kwargs)
+        return self.plates[name]
+
+    def group(self, match=".*"):
+        """
+        Select a :class:`Group` of model sites for joint guidance.
+
+        :param str match: A regex string matching names of model sample sites.
+        :return: A group of model sites.
+        :rtype: Group
+        """
+        if match not in self.groups:
+            sites = [site
+                     for name, site in self.prototype_trace.iter_stochastic_nodes()
+                     if re.match(match, name)]
+            if not sites:
+                raise ValueError("EasyGuide.group() pattern {} matched no model sites"
+                                 .format(repr(match)))
+            self.groups[match] = Group(self, sites)
+        return self.groups[match]
+
+    def map_estimate(self, name):
+        """
+        Construct a maximum a posteriori (MAP) guide using Delta distributions.
+
+        :param str name: The name of a model sample site.
+        :return: A sampled value.
+        :rtype: torch.Tensor
+        """
+        site = self.prototype_trace.nodes[name]
+        fn = site["fn"]
+        param_name = "auto_{}".format(name)
+        init_value = None
+        if param_name not in pyro.get_param_store():
+            init_value = site["value"].detach()
+        with ExitStack() as stack:
+            for frame in site["cond_indep_stack"]:
+                plate = self.plate(frame.name)
+                if plate not in runtime._PYRO_STACK:
+                    stack.enter_context(plate)
+                elif init_value is not None and plate.subsample_size < plate.size:
+                    # Repeat the init_value to full size.
+                    dim = plate.dim - fn.event_dim
+                    assert init_value.size(dim) == plate.subsample_size
+                    ind = torch.arange(plate.size, device=init_value.device)
+                    ind = ind % plate.subsample_size
+                    init_value = init_value.index_select(dim, ind)
+            value = pyro.param(param_name, init_value,
+                               constraint=fn.support, event_dim=fn.event_dim)
+            return pyro.sample(name, dist.Delta(value, event_dim=fn.event_dim))
+
+
+class Group(object):
+    """
+    An autoguide helper to match a group of model sites.
+
+    :ivar list sites: A list of all matching sample sites in the model.
+    :ivar torch.Size event_shape: The total flattened concatenated shape of all
+        matching sample sites in the model.
+    :param EasyGuide guide: An easyguide instance.
+    :param list sites: A list of model sites.
+    """
+    def __init__(self, guide, sites):
+        assert isinstance(sites, list)
+        assert sites
+        self._guide = weakref.ref(guide)
+        self.sites = sites
+
+        # A group is in a frame only if all its sample sites are in that frame.
```

Comment:
what happens in 176ff if a group of sites crosses a frame boundary?

Reply:
Then `self.frames` will be set to only those frames populated by all `sites`. If a frame is occupied by a strict subset of `sites` (and hence not in `self.frames`) then than frame cannot be subsampled (if it is subsampled, an error will be raised).

Reply:
where exactly does that error get triggered?

Reply:
Hmm good point. I've added an informative error message, as well as two new tests:
1. test that it's ok to mix shared and non-shared plates if all shared plates are left of non-shared plates.
2. test that the error is triggered if a shared plate is right of a non-shared plate.

Reply:
thanks for the additional tests! i'm still a bit confused about this. would like to discuss offline tomorrow before merging